src/mapport.cpp

```diff
@@ -0,0 +1,342 @@
+        natpmpresp_t response;
+        do {
+            r_read = readnatpmpresponseorretry(natpmp, &response);
+        } while (r_read == NATPMP_TRYAGAIN);
+
+        if (r_read == 0) {
+            external_ipv4_addr = response.pnu.publicaddress.addr;
+            return true;
+        } else if (r_read == NATPMP_ERR_NOGATEWAYSUPPORT) {
+            LogPrintf("NAT-PMP: The gateway does not support NAT-PMP.\n");
+        } else {
+            LogPrintf("NAT-PMP: readnatpmpresponseorretry() for public address failed with %d error.\n", r_read);
+        }
+
+    } else {
+        LogPrintf("NAT-PMP: sendpublicaddressrequest() failed with %d error.\n", r_send);
+    }
+
+    return false;
+}
+
+static bool NatpmpMapping(natpmp_t* natpmp, const struct in_addr& external_ipv4_addr, uint16_t private_port, bool& external_ip_discovered)
+{
+    const uint16_t suggested_external_port = g_mapport_external_port ? g_mapport_external_port : private_port;
+    const int r_send = sendnewportmappingrequest(natpmp, NATPMP_PROTOCOL_TCP, private_port, suggested_external_port, 3600 /*seconds*/);
+    if (r_send == 12 /* OK */) {
+        int r_read;
+        natpmpresp_t response;
+        do {
+            r_read = readnatpmpresponseorretry(natpmp, &response);
+        } while (r_read == NATPMP_TRYAGAIN);
+
+        if (r_read == 0) {
+            auto pm = response.pnu.newportmapping;
+            if (private_port == pm.privateport && pm.lifetime > 0) {
+                g_mapport_external_port = pm.mappedpublicport;
+                const CService external{external_ipv4_addr, pm.mappedpublicport};
+                if (!external_ip_discovered && fDiscover) {
+                    AddLocal(external, LOCAL_MAPPED);
+                    external_ip_discovered = true;
+                }
+
+                LogPrintf("NAT-PMP: Port mapping successful. External address = %s\n", external.ToString());
+                return true;
+            } else {
+                LogPrintf("NAT-PMP: Port mapping failed.\n");
+            }
+
+        } else if (r_read == NATPMP_ERR_NOGATEWAYSUPPORT) {
+            LogPrintf("NAT-PMP: The gateway does not support NAT-PMP.\n");
+        } else {
+            LogPrintf("NAT-PMP: readnatpmpresponseorretry() for port mapping failed with %d error.\n", r_read);
+        }
+
+    } else {
+        LogPrintf("NAT-PMP: sendnewportmappingrequest() failed with %d error.\n", r_send);
+    }
+
+    return false;
+}
+
+static bool ThreadNatpmp()
+{
+    bool ret = false;
+    natpmp_t natpmp;
+    struct in_addr external_ipv4_addr;
+    if (NatpmpInit(&natpmp) && NatpmpDiscover(&natpmp, external_ipv4_addr)) {
+        bool external_ip_discovered = false;
+        const uint16_t private_port = GetListenPort();
+        do {
+            ret = NatpmpMapping(&natpmp, external_ipv4_addr, private_port, external_ip_discovered);
+        } while (ret && g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+        g_mapport_interrupt.reset();
```

Comment:
@luke-jr 
> This should be reset outside of the protocol-specific functions, or else we could end up with a race.

Looking into the `CThreadInterrupt` class implementation I couldn't see possibilities for a race.

What did I miss?